### ISSUE TITLE
Templates don't work for __construct params

### ISSUE THREAD

**Issue by iluuu1994** [vimeo/psalm#1585]
Not sure if this is a known issue:

```php
/**
 * @template T
 */
class Foo {
  /**
   * @param T $foo
   */
  public function __construct($foo)
  {
  }
  
  /**
   * @param T $foo
   */
  public function setFoo($foo): void
  {
  }
}

/** @var Foo<string> $foo */
$foo = new Foo(10); // <- No warning
$foo->setFoo(10); // <- Warning
```

https://psalm.dev/r/a4a2428aae

Psalm does not warn when a generic `__construct` parameter doesn't match the given type.

**Comment by weirdan** [comment]
Actually what you're seeing here boils down to the fact that Psalm doesn't warn on the assignments having var docblock at all: https://psalm.dev/r/2ee7b85fa9

**Comment by muglug** [comment]
Yeah, and that's by design - `@var` docblocks override whatever's assigned below them

**Comment by iluuu1994** [comment]
Oh sorry guys, that makes sense.

**Comment by muglug** [comment]
Here's a slightly more real-world example: https://psalm.dev/r/351473eff2

Passing the parameter binds the object on creation.

**Comment by iluuu1994** [comment]
@muglug [This example](https://psalm.dev/r/afb63475f5) shows that the constructor params are validated. My bad.

**Comment by muglug** [comment]
@iluuu1994 no worries - there are [a lot of tests](https://github.com/vimeo/psalm/tree/master/tests/Template) that should cover this, and much more functionality
